src/monitoring.c

```diff
@@ -314,43 +562,137 @@ static void handle_client(struct monitoring *monitoring, int fd)
 static void *monitoring_thread(void * p_data)
 {
 	struct monitoring *monitoring;
-	struct sockaddr_in client_addr;
-	socklen_t length;
-	int fd, ret;
 	bool stop;
 
 	monitoring = (struct monitoring*) p_data;
 	stop = monitoring->stop;
-	fd_set readfds;
+
+	int epollfd = epoll_create1(0);
+	if (epollfd < 0) {
+		log_error("epoll_create1");
+		return NULL;
+	}
+
+	struct epoll_event accept_event;
+	accept_event.data.fd = monitoring->sockfd;
+	accept_event.events = EPOLLIN;
+	if (epoll_ctl(epollfd, EPOLL_CTL_ADD, monitoring->sockfd, &accept_event) < 0) {
+		log_error("epoll_ctl EPOLL_CTL_ADD");
+		return NULL;
+	}
+
+	struct epoll_event* events = calloc(MAXFDS, sizeof(struct epoll_event));
+	if (events == NULL) {
+		log_error("Unable to allocate memory for epoll_events");
+		return NULL;
+	}
 
 	while (!stop)
 	{
 		log_trace("Monitoring: Listening on socket...");
-		ret = listen(monitoring->sockfd, 1);
-		if (ret == -1) {
-			log_error("Monitoring: Error listening to socket");
-			return NULL;
-		}
-
-		FD_ZERO(&readfds);
-		FD_SET(monitoring->sockfd, &readfds);
-		struct timeval tv = {SOCKET_TIMEOUT, 0};
-		int sel = select(monitoring->sockfd + 1, &readfds, 0, 0, &tv);
-		if (sel > 0) {
-			length = sizeof(client_addr);
-			fd = accept(monitoring->sockfd, (struct sockaddr *)&client_addr, &length);
-			if (fd == -1) {
-				log_error("Monitoring: Error accepting client connection");
-				continue;
+		int nready = epoll_wait(epollfd, events, MAXFDS, SOCKET_TIMEOUT_MS);
+		for (int i = 0; i < nready; i++) {
+			if (events[i].events & EPOLLERR) {
+				log_error("epoll_wait returned EPOLLERR");
+				return NULL;
 			}
 
-			if (fd > 0) {
-				handle_client(monitoring, fd);
+			if (events[i].data.fd == monitoring->sockfd) {
+				// The listening socket is ready; this means a new peer is connecting.
+
+				struct sockaddr_in peer_addr;
+				socklen_t peer_addr_len = sizeof(peer_addr);
+				int newsockfd = accept(monitoring->sockfd, (struct sockaddr*)&peer_addr,
+									&peer_addr_len);
+				if (newsockfd < 0) {
+				if (errno == EAGAIN || errno == EWOULDBLOCK) {
+					// This can happen due to the nonblocking socket mode; in this
+					// case don't do anything, but print a notice (since these events
+					// are extremely rare and interesting to observe...)
+					log_debug("accept returned EAGAIN or EWOULDBLOCK");
+				} else {
+					log_error("accept");
+					return NULL;
```

Comment:
To be clear, newsckfd < 0 means new connexion to be accepted, newsockfd >= 0 means already accepted and to be configurated ?

Reply:
![Screenshot_20220513_115138](https://user-images.githubusercontent.com/6388268/168259053-9555ca3e-937b-4824-959f-68d0abcfb09e.png)
Negative values means an error ocurred when opening the socket, positive values means that we have been ableto open the socket